src/index.js

```diff
@@ -1,47 +1,94 @@
-/** @license MIT License (c) copyright 2015-2017 original author or authors */
-
 import { MulticastSource } from '@most/core'
+import { asap, cancelTask } from '@most/scheduler'
 
-// hold :: Stream a -> Stream a
 export const hold = stream =>
-  new MulticastSource(new Hold(stream))
+  new Hold(stream)
 
-class Hold {
+export class Hold extends MulticastSource {
   constructor (source) {
-    this.source = source
-    this.time = -Infinity
-    this.value = void 0
+    super(source)
+    this.hasValue = false
+    this.value = undefined
+    this.pendingSinks = []
+    this.task = undefined
   }
 
   run (sink, scheduler) {
-    /* istanbul ignore else */
-    if (sink._hold !== this) {
-      sink._hold = this
-      sink._holdAdd = sink.add
-      sink.add = holdAdd
-
-      sink._holdEvent = sink.event
-      sink.event = holdEvent
+    if (this.hasValue && this.sinks.length > 0) {
+      this._scheduleFlush(sink, scheduler)
+    }
+
+    return super.run(sink, scheduler)
+  }
+
+  dispose () {
+    this._cancelTask()
+    return super.dispose()
+  }
+
+  event (time, value) {
+    this.hasValue = true
+    this.value = value
+    const pendingSinks = this._flushPending(time)
+    super.event(time, value)
+    this.sinks = this.sinks.concat(pendingSinks)
+  }
+
+  _flushPending (time) {
+    const pendingSinks = this.pendingSinks
+    this.pendingSinks = []
+    for (let i = 0; i < pendingSinks.length; ++i) {
+      tryEvent(time, this.value, pendingSinks[i])
     }
+    return pendingSinks
+  }
 
-    return this.source.run(sink, scheduler)
+  _scheduleFlush (sink, scheduler) {
+    this.pendingSinks.push(sink)
+    if (!this.task) {
+      cancelTask(this.task)
```

Comment:
Am I missing something? `cancelTask` takes a `scheduledTask` and disposes of it. However, the if-statement only allows `this.task` to be `undefined`, `false`, `null`, `0`, or `NaN`. How is it possible to cancel these values?